libs/wingc/src/type_check.rs

```diff
@@ -2584,6 +2596,49 @@ impl<'a> TypeChecker<'a> {
 		}
 	}
 
+	/// Validate that the given type is a subtype (or same) as the expected type, disregarding whether
+	/// the object is mutable or immutable. If not, add an error to the diagnostics.
+	///
+	/// Returns the given type on success, otherwise returns the expected type.
+	fn validate_type_binary_operator(
+		&mut self,
+		actual_type: TypeRef,
+		expected_type: TypeRef,
+		span: &impl Spanned,
+	) -> TypeRef {
+		if actual_type.is_array_type()
+			&& expected_type.is_array_type()
```

Comment:
1. My hunch is that using pattern matching would be cleaner here. 
2. We don't need to rely on the inner type being the exact same, we should keep using validate_type because we can use `==` with subtypes.
3. If we're validating the inner type, this function should be recursive since you can do `MutArray<num>[Array<num>[1]] == Array<num>[Array<num>[1]]`

So maybe something like this?

```rust
		if let Type::Array(inner_actual) | Type::MutArray(inner_actual) = *actual_type {
			if let Type::Array(inner_expected) | Type::MutArray(inner_expected) = *expected_type {
				return self.validate_type_binary_operator(inner_actual, inner_expected, span);
			}
		}
```

If using this, the new `is_*_type()` functions can be removed since they'll be unused.

Reply:
Another option, with less nesting (although requiring an extra dereferencing):

```rust
		if let (
			Type::Array(inner_actual) | Type::MutArray(inner_actual),
			Type::Array(inner_expected) | Type::MutArray(inner_expected),
		) = (&*actual_type, &*expected_type)
		{
			return self.validate_type_binary_operator(*inner_actual, *inner_expected, span);
		}
```